src/Foundatio.Storage.SshNet/Storage/SshNetFileStorage.cs

```diff
@@ -13,384 +14,432 @@
-                hasMore = true;
-                list.RemoveAt(pagingLimit - 1);
-            }
+    private async Task<List<FileSpec>> GetFileListAsync(string searchPattern = null, int? limit = null, int? skip = null, CancellationToken cancellationToken = default) {
+        using var _ = _logger.BeginScope(s => s.Property("SearchPattern", searchPattern).Property("Limit", limit).Property("Skip", skip));
+        
+        if (limit is <= 0)
+            return new List<FileSpec>();
 
-            return new NextPageResult {
-                Success = true,
-                HasMore = hasMore,
-                Files = list,
-                NextPageFunc = hasMore ? r => GetFiles(searchPattern, page + 1, pageSize, cancellationToken) : null
-            };
-        }
+        var list = new List<FileSpec>();
+        var criteria = GetRequestCriteria(searchPattern);
 
-        private async Task<List<FileSpec>> GetFileListAsync(string searchPattern = null, int? limit = null, int? skip = null, CancellationToken cancellationToken = default) {
-            if (limit is <= 0)
-                return new List<FileSpec>();
+        EnsureClientConnected();
 
-            var list = new List<FileSpec>();
-            var criteria = GetRequestCriteria(searchPattern);
+        // NOTE: This could be very expensive the larger the directory structure you have as we aren't efficiently doing paging.
+        int? recordsToReturn = limit.HasValue ? skip.GetValueOrDefault() * limit + limit : null;
+        await GetFileListRecursivelyAsync(criteria.Prefix, criteria.Pattern, list, recordsToReturn, cancellationToken).AnyContext();
 
-            EnsureClientConnected();
+        if (skip.HasValue)
+            list = list.Skip(skip.Value).ToList();
 
-            // NOTE: This could be very expensive the larger the directory structure you have as we aren't efficiently doing paging.
-            int? recordsToReturn = limit.HasValue ? skip.GetValueOrDefault() * limit + limit : null;
-            await GetFileListRecursivelyAsync(criteria.Prefix, criteria.Pattern, list, recordsToReturn, cancellationToken).AnyContext();
+        if (limit.HasValue)
+            list = list.Take(limit.Value).ToList();
 
-            if (skip.HasValue)
-                list = list.Skip(skip.Value).ToList();
+        return list;
+    }
 
-            if (limit.HasValue)
-                list = list.Take(limit.Value).ToList();
+    private async Task GetFileListRecursivelyAsync(string prefix, Regex pattern, ICollection<FileSpec> list, int? recordsToReturn = null, CancellationToken cancellationToken = default) {
+        using var _ = _logger.BeginScope(s => s.Property("Prefix", prefix).Property("Pattern", pattern.ToString()));
 
-            return list;
+        _logger.LogTrace("Getting file list recursively matching {Prefix}...", prefix);
+        if (cancellationToken.IsCancellationRequested) {
+            _logger.LogDebug("Returning empty file list: Cancellation requested");
```

Comment:
I would just say cancellation requested. This method is called recursively so it might have already returned a bunch of results.

Reply:
You have no context when reading the log message if all it says is Cancellation requested

Reply:
Well saying that you are returning an empty file list because the request was cancelled is even worse because you are saying something that might not be true.

Reply:
It's a recursive call with the parent saying total count returned.

Reply:
Right, and the recursion is all building up the same collection. So if it exits from cancellation after it's already got data then the resulting list will have data in it.